jetty-core/jetty-server/src/main/java/org/eclipse/jetty/server/handler/ConditionalHandler.java

```diff
@@ -0,0 +1,659 @@
+     */
+    public static class InetAddressPredicate implements Predicate<Request>
+    {
+        public static InetAddress getInetAddress(SocketAddress socketAddress)
+        {
+            if (socketAddress instanceof InetSocketAddress inetSocketAddress)
+            {
+                if (inetSocketAddress.isUnresolved())
+                {
+                    try
+                    {
+                        return InetAddress.getByName(inetSocketAddress.getHostString());
+                    }
+                    catch (UnknownHostException e)
+                    {
+                        if (LOG.isTraceEnabled())
+                            LOG.trace("ignored", e);
+                        return null;
+                    }
+                }
+
+                return inetSocketAddress.getAddress();
+            }
+            return null;
+        }
+
+        private final InetAddressPattern _pattern;
+
+        public InetAddressPredicate(InetAddressPattern pattern)
+        {
+            _pattern = pattern;
+        }
+
+        @Override
+        public boolean test(Request request)
+        {
+            return _pattern.test(getInetAddress(request.getConnectionMetaData().getRemoteSocketAddress()));
+        }
+
+        @Override
+        public int hashCode()
+        {
+            return _pattern.hashCode();
+        }
+
+        @Override
+        public boolean equals(Object obj)
+        {
+            return obj instanceof InetAddressPredicate inetAddressPredicate && _pattern.equals(inetAddressPredicate._pattern);
+        }
+
+        @Override
+        public String toString()
+        {
+            return super.toString();
+        }
+    }
+
+    /**
+     * A {@link Predicate} over {@link Request} that tests on or more of:
+     * <ul>
+     *     <li>The {@link Connector#getName() name} of the
+     *     {@link ConnectionMetaData#getConnector() connector}
+     *     obtained from {@link Request#getConnectionMetaData()}</li>
+     *     <li>An {@link InetAddressPattern} tested against the
+     *     {@link ConnectionMetaData#getRemoteSocketAddress() getRemoteSocketAddress()} of
+     *     {@link Request#getConnectionMetaData()}.
+     *     </li>
+     *     <li>A {@link Request#getMethod() method} name</li>
+     *     <li>A {@link PathSpec} tested against the {@link Request#getPathInContext(Request) pathInContext}.</li>
+     * </ul>
+     */
+    public static class ConnectorAddrMethodPathPredicate implements Predicate<Request>
+    {
+        private final String connector;
+        private final InetAddressPattern address;
+        private final String method;
+        private final PathSpec pathSpec;
+
+        public ConnectorAddrMethodPathPredicate(String connector, InetAddressPattern address, String method, PathSpec pathSpec)
+        {
+            this.connector = connector;
+            this.address = address;
+            this.method = method;
+            this.pathSpec = pathSpec;
```

Comment:
Use the `Predicate` system properly and you won't need these uber / composite classes.

The individual tests here (connector name, http method name, path spec, inet address) should be separate predicates.

Then combined with `Predicate.and()`, `Predicate.or()`, `Predicate.not()`, et-al techniques.

Reply:
Good  idea. I'll give it a try.

Reply:
@joakime that actually was a big improvement and simplification.   It has also allowed me to move the optimizations from utility methods to the general include/exclude predicate API.